flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -107,48 +127,74 @@ public Boundedness getBoundedness() {
     }
 
     /**
-     * Converts checkpoint between sources to transfer end position to next source's start position.
-     * Only required for dynamic position transfer at time of switching, otherwise source can be
+     * Callback for switch time customization of the underlying source from previous enumerator end
+     * state.
+     *
+     * <p>Called when the current enumerator has finished and before the next enumerator is created.
+     * The enumerator end state can thus be used to set the next source's start start position.
+     *
+     * <p>Only required for dynamic position transfer at time of switching, otherwise source can be
      * preconfigured with a start position during job submission.
      */
-    public interface CheckpointConverter<InCheckpointT, OutCheckpointT>
-            extends Function<InCheckpointT, OutCheckpointT>, Serializable {}
+    public interface SourceConfigurer<SourceT extends Source, FromEnumT extends SplitEnumerator>
+            extends Serializable {
+        SourceT configure(SourceT source, FromEnumT enumerator);
+    }
+
+    private static class NoopSourceConfigurer<
+                    SourceT extends Source, FromEnumT extends SplitEnumerator>
+            implements SourceConfigurer<SourceT, FromEnumT> {
+        @Override
+        public SourceT configure(SourceT source, FromEnumT enumerator) {
+            return source;
+        }
+    }
 
-    /** Chain of sources with option to convert start position at switch-time. */
-    public static class SourceChain<T, EnumChkT> implements Serializable {
-        final List<Tuple2<Source<T, ? extends SourceSplit, ?>, CheckpointConverter<?, ?>>> sources;
+    /** Entry for list of underlying sources. */
+    protected static class SourceListEntry implements Serializable {
+        protected final Source source;
+        protected final SourceConfigurer configurer;
 
-        public SourceChain(Source<T, ?, EnumChkT> initialSource) {
-            this(concat(Collections.emptyList(), Tuple2.of(initialSource, null)));
+        private SourceListEntry(Source source, SourceConfigurer configurer) {
+            this.source = source;
+            this.configurer = configurer;
```

Comment:
`checkNotNull`, we just force the `NoopSourceConfigurer` where we used `null` before.